cmd/registry/cmd/upload/bulk/protos.go

```diff
@@ -117,127 +116,111 @@ func (task *uploadProtoTask) Run(ctx context.Context) error {
-	if _, err := task.client.GetApiVersion(ctx, &rpc.GetApiVersionRequest{
-		Name: task.versionName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when version is found without error.
-	}
-
-	response, err := task.client.CreateApiVersion(ctx, &rpc.CreateApiVersionRequest{
-		Parent:       task.apiName(),
-		ApiVersionId: task.versionID,
-		ApiVersion:   &rpc.ApiVersion{},
+	// Create an API version if needed (or update an existing one)
+	response, err := task.client.UpdateApiVersion(ctx, &rpc.UpdateApiVersionRequest{
+		ApiVersion: &rpc.ApiVersion{
+			Name: task.versionName(),
+		},
+		AllowMissing: true,
 	})
-	if err != nil {
-		log.WithError(err).Debugf("Failed to create version %s", task.versionName())
+	if err == nil {
+		log.Debugf("Updated %s", response.Name)
 	} else {
-		log.Debugf("Created %s", response.Name)
+		log.WithError(err).Debugf("Failed to create version %s", task.versionName())
 	}
 
 	return nil
 }
 
-func (task *uploadProtoTask) createSpec(ctx context.Context) error {
+func (task *uploadProtoTask) createOrUpdateSpec(ctx context.Context) error {
 	contents, err := task.zipContents()
 	if err != nil {
 		return err
 	}
 
-	if _, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
+	// Use the spec size and hash to avoid unnecessary uploads.
+	spec, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
 		Name: task.specName(),
-	}); !core.NotFound(err) {
-		return err // Returns nil when spec is found without error.
-	}
-
-	request := &rpc.CreateApiSpecRequest{
-		Parent:    task.versionName(),
-		ApiSpecId: task.fileName(),
-		ApiSpec: &rpc.ApiSpec{
-			MimeType: core.ProtobufMimeType("+zip"),
-			Filename: task.fileName(),
-			Contents: contents,
-		},
-	}
-	if task.baseURI != "" {
-		request.ApiSpec.SourceUri = fmt.Sprintf("%s/%s", task.baseURI, task.apiPath())
-	}
-
-	response, err := task.client.CreateApiSpec(ctx, request)
-	if err != nil {
-		log.WithError(err).Errorf("Error %s [contents-length: %d]", task.specName(), len(contents))
-	} else {
-		log.Debugf("Created %s", response.Name)
-	}
-
-	return nil
-}
+	})
 
-func (task *uploadProtoTask) updateSpec(ctx context.Context) error {
-	contents, err := task.zipContents()
-	if err != nil {
-		return err
+	if err == nil && int(spec.GetSizeBytes()) == len(contents) && spec.GetHash() == hashForBytes(contents) {
+		log.Debugf("Matched already uploaded spec %s", task.specName())
+		return nil
 	}
 
-	spec, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
+	// Use the spec size and hash to avoid unnecessary uploads.
+	if spec, err := task.client.GetApiSpec(ctx, &rpc.GetApiSpecRequest{
 		Name: task.specName(),
-	})
-	if err != nil && !core.NotFound(err) {
-		return err
+	}); err == nil {
+		if int(spec.GetSizeBytes()) == len(contents) {
+			hash := hashForBytes(contents)
+			if spec.GetHash() == hash {
+				log.Debugf("Matched already uploaded spec %s", task.specName())
+				return nil
+			}
+		}
```

Comment:
Shouldn't this be removed now that the flattened check is included above?

Reply:
Yes, I missed this. Thanks for pointing this out! (removed now)